src/poetry/repositories/pool.py

```diff
@@ -1,59 +1,77 @@
 from __future__ import annotations
 
+from collections import OrderedDict
+from dataclasses import dataclass
+from enum import IntEnum
+from enum import auto
 from typing import TYPE_CHECKING
 
+from poetry.repositories.abstract_repository import AbstractRepository
 from poetry.repositories.exceptions import PackageNotFound
-from poetry.repositories.repository import Repository
 
 
 if TYPE_CHECKING:
     from poetry.core.constraints.version import Version
     from poetry.core.packages.dependency import Dependency
     from poetry.core.packages.package import Package
 
+    from poetry.repositories.repository import Repository
 
-class Pool(Repository):
+
+class RepositoryPriority(IntEnum):
+    # The order of the members below dictates the actual priority. The first member has
+    # top priority.
+    DEFAULT = auto()
+    PRIMARY = auto()
+    SECONDARY = auto()
+
+
+@dataclass
+class PrioritizedRepository:
+    repository: Repository
+    priority: RepositoryPriority
+
+
+class Pool(AbstractRepository):
     def __init__(
         self,
         repositories: list[Repository] | None = None,
         ignore_repository_names: bool = False,
     ) -> None:
-        super().__init__("poetry-pool")
-
-        if repositories is None:
-            repositories = []
-
-        self._lookup: dict[str, int] = {}
-        self._repositories: list[Repository] = []
-        self._default = False
-        self._has_primary_repositories = False
-        self._secondary_start_idx: int | None = None
+        super().__init__("poetry-repository-pool")
+        self._repositories: OrderedDict[str, PrioritizedRepository] = OrderedDict()
+        self._ignore_repository_names = ignore_repository_names
 
+        repositories = repositories or []
```

Comment:
This is pretty messy -- it doesn't look like we ever call `Pool(None)`, so it should be perfectly safe to change this to a default value without any other code changes.

Reply:
We do call `Pool()` in many places. Which default value are you thinking of (assuming you don't want to have a mutable object as default)?

Reply:
Ah, shoot, I apparently have spent too much time in other languages (*cough*Ruby*cough*) -- I appear to have forgotten about mutable default values last night. My bad :laughing: